### PR TITLE
Refactor `Ouroboros.Network.Diffusion.runM`

### LINKED ISSUE #3834: Refactor `Ouroboro.Network.Diffusion.runM`
* The large amounts of code duplication makes it challenging to understand runM.
* It is likely that one could write `runM` more concisely as well as more clearly. As this function is long and inherently complicated, this could be a good investment of effort.

### PR DESCRIPTION
**Pull request by mtullsen**
# Description

resolves #3834 : Refactor `Ouroboros.Network.Diffusion.runM`.

## Motivation

`runM` (called by `run`) is 640 lines of code, and is currently not very easy to follow and there is a significant amount of code duplication.
There are two primary case alternatives in runM, where currently we have a lot of duplication; abstractly:
```
  case diffusionMode of
   InitiatorOnlyDiffusionMode -> ...
   InitiatorAndResponderDiffusionMode -> ...
```

## Objectives of refactor

Our objectives here are
- reduce linecount, if possible
- more importantly, remove code duplication
- restructure code so as to have 80 character lines and no "indentation fudging".
- make the similarities and differences between the two diffusion modes more apparent.
- make the "primary control flow" of the two modes readily apparent (and comparable to one another).

## Notes to reviewers
  
- This "big refactor" is a sequence of a couple dozen small refactors, each which should be relatively obvious.
- Not every refactor may seem to be compelling; but most all will enable future refactors.  For justification/motivation, one may want to view the final result.
- Many of these "refactors" involve large moves where git diff gives poor results; but one might, using the command-line, view some of these with something such as
  ```
  git show -w --color-moved=dimmed-zebra --color-moved-ws=ignore-space-change <COMMIT>
  ```

# Checklist

- Branch
    - [x] Updated changelog files.
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] The documentation has been properly updated
    - [x] New tests are added if needed and existing tests are updated
    - [x] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

**Review by coot — APPROVED** [review]
This looks great.  You were absolutely right I will like the PR.

Maybe we should move the specialised functions to `Ouroboros.Network.Diffusion.P2P.{NodeToClient,NodeToNode}` modules, then `runM`?

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 1]
```diff
@@ -846,6 +807,77 @@ runM Interfaces
...
+              dtMuxTracer
+              muxMode
+              diNtnHandshakeArguments
+              versions
+              (mainThreadId, rethrowPolicy <> daRethrowPolicy)
+
+          -- Capture the two variations (InitiatorMode,InitiatorResponderMode) of
+          -- withConnectionManager:
+          
+          iomWithConnectionManager =
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs]: This was addressed in a later commit.

**Comment by coot** [comment]
`haskell-stylish` complains in CI.  You can run `./script/ci/check-stylish` if you have `stylish-haskell-0.14.4.0` installed on your system or run `nix build .\#stylish-haskell`.

**Comment by coot** [comment]
There is a conflict with master, it's due to 8fa5913ba7.

**Comment by mtullsen** [comment]
@coot fixed merge conflict

**Pull request merged**